### ISSUE TITLE
Best way to access the store?

### ISSUE THREAD

**Issue by sospedra** [yelouafi/redux-saga#640]
Imagine we've the following state:

```js
{
  user: 'username',
  count: 0,
  token: 'secretTokenGivenAfterAuth'
}
```

And **our API needs the token** as a query parameter (under HTTPS) **on every single request**. The more straight forward implementation would be something like:

*Note: the following code is a lazy implementation*

```js
import * as API from './api'

const fetchSomething = function * (action) {
  const { endpoint, body, token } = action
  const payload = yield call(API.makeRequest, endpoint, body, token)
  yield put(fetchSomethingSucceed(payload))
}

export default function * () {
  yield take(FETCH_SOMETHING, fetchSomething)
}
```

But here's a strict contract: **we need to send the token on every action payload**. Which feels not DRY and prone to errors.

Is there any best practice to **connect the store to the Saga**? To enhance each fetch request with the token; saved on the sore. After all is a middleware and `store.getState` is provided by Redux.

Many thanks in advance :D

**Comment by Andarist** [comment]
you can access the `store` with `select` effect

```javascript
const getToken = (state) => state.token
// ...
const token = yield select(getToken)
```

**Comment by sospedra** [comment]
Awesome @Andarist ! This is exactly what I was looking for. I've just read the documentation and found the note about don't relying on the Store state. The saga being autonomous. However, there are excepctions. I think that requesting a token is something which fulfill to break this rule. Maybe could be an example. How do you feel about this? Maybe I'm completely wrong :)

**Comment by Andarist** [comment]
Well, it depends. Personally I try to avoid accessing the store's state (as mentioned in the docs) unless I dont replicate reducers logic of parsing actions and updating the saga's state - in such case its simply not worth it.

But for your use case something similiar to this could be used to avoid `select`

```javascript
import * as API from './api'

const fetchSomething = function * (token, action) {
  const { endpoint, body } = action
  const payload = yield call(API.makeRequest, endpoint, body, token)
  yield put(fetchSomethingSucceed(payload))
}

export default function * () {
  const { token } = yield take(SET_TOKEN)
  yield take(FETCH_SOMETHING, fetchSomething, token)
}
```

I assume you have some action which sets the token in the store already.

**Comment by sospedra** [comment]
It sounds good. But I'm facing another issue because `take` is bloquer. Therefore, `FETCH_SOMETHING` will be called only after 'SET_TOKEN`. The requested flow is like:

```
>> FETCH_SOMETHING >> SET_TOKEN >> FETCH_SOMETHING
```

Which means that `FETCH_SOMETHING` is a watcher which can **has or not** the token.

I tried with `takeEvery` but it didn't worked. Any suggestion?
Many thanks, again.

**Comment by Andarist** [comment]
Just from top of my head, smth like this should work (although there might be a better way)
```javascript
import * as API from './api'

const fetchSomething = function * ({ token }, action) {
  const { endpoint, body } = action
  const payload = yield call(API.makeRequest, endpoint, body, token)
  yield put(fetchSomethingSucceed(payload))
}

export default function * () {
  let token

  yield fork(function* () {
   token = (yield take(SET_TOKEN)).token
  })
  while (true) {
    yield take(FETCH_SOMETHING, fetchSomething, { token })
  }
}
```

**Comment by sospedra** [comment]
Many many thanks @Andarist !
Last step I solved with `takeEvery` and now everything is working nice and clean!
I'm closing the issue, tho, I think an example (or receipt) with a flow like this one would be super cool ;)

**Comment by Ayyagaries** [comment]
I am in the same situation, here I have two sagas one is login that gives me the token, and other one is user profile that gives me the profile with the token. 
I used  take("GET_USER_TOKEN") in the user profile saga to get the token but I noticed that the action is called twice in the "Login saga" and the 'user profile saga'. I want to action to happen only once. How is that possible?

**Comment by BenLorantfy** [comment]
@Andarist Why is that solution better than just selecting from the state? It seems like you're just creating a local state inside your saga. Isn't redux meant to consolidate state in one spot?

**Comment by Andarist** [comment]
@BenLorantfy I've proposed to use `select` and only provided this more contrived example to illustrate that possibility. I would just go with `select` for this.

**Comment by iorrah** [comment]
For those who are wondering where the `select` comes from:

```
import { select } from 'redux-saga/effects';
const getToken = (state) => state.token;
// ...
const token = yield select(getToken);
```

**Comment by PabloDinella** [comment]
> For those who are wondering where the `select` comes from:
> 
> ```js-jsx
> import { select } from 'redux-saga/effects';    //    <- here it is
> 
> const getToken = (state) => state.token;
> // ...
> const token = yield select(getToken);
> ```

and here is the doc section about it :)
https://redux-saga.js.org/docs/api/#selectselector-args

**Comment by vipulrawat** [comment]
Ok, one thing: is that a good practice to store authentication token to the client side store?

**Comment by ufon** [comment]
@vipulrawat why not

**Comment by vipulrawat** [comment]
@ufon Wouldn't that be a security concern in storing the authentication token in the client side?

**Comment by BenLorantfy** [comment]
@vipulrawat I think it's fine. You have to store the client side token somewhere, and the redux store isn't too much different than localStorage. The only concern I'd have is making sure you don't accidentally send the token to an analytics or logging service. 

You may also be interested in `actionSanitizer` and `stateSanitizer` to remove the token from the devTools: https://github.com/zalmoxisus/redux-devtools-extension/blob/master/docs/API/Arguments.md#actionsanitizer--statesanitizer
